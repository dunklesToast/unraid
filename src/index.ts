export * as ssh from './executors/SSH';
export * as vm from './modules/vms/vm-module';
