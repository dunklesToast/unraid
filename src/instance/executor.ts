import type { EventEmitter } from 'events';

export type IExecute = {
    command: string;
};

export type IExecuteResult = {
    stdout: string[];
    stderr: string[];
    code: number;
    signal?: unknown;
};

export type IExecuteStreamResult = {
    code: number;
    signal?: unknown;
};

export type IExecuteSimple = string;

export type CancelFunction = () => Promise<void>;

export type ExecutorConfigType = Record<string, unknown>;

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export interface Executor<Config> {
    executeStream?(
        command: IExecuteSimple | IExecute
    ): Promise<[EventEmitter, CancelFunction, Promise<IExecuteStreamResult>]>;
}

export abstract class Executor<Config extends ExecutorConfigType = ExecutorConfigType> {
    readonly config: Config;

    constructor(config: Config) {
        this.config = config;
    }

    abstract execute(command: IExecuteSimple): Promise<IExecuteResult>;

    abstract execute({ command }: IExecute): Promise<IExecuteResult>;

    abstract execute(command: IExecuteSimple | IExecute): Promise<IExecuteResult>;
}
