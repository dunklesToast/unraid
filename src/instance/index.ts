export * as executor from './executor';
export * as unraid from './unraid';
