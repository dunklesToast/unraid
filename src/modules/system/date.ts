import { execute } from '../../executors/SSH';

/**
 * Returns the current Server Datetime
 * @param parsed - If the date should be returned as JS Date Object
 */
async function date(parsed?: false): Promise<string>;
async function date(parsed?: true): Promise<Date>;
async function date(parsed?: boolean): Promise<string | Date> {
  const { code, stdout } = await execute(`date${parsed ? ' -R' : ''}`);
  if (code !== 0) throw new Error('Got non-zero exit code while getting date');
  if (parsed) {
    return new Date(stdout.join(''));
  }
  return stdout.join('');
}

export { date };
