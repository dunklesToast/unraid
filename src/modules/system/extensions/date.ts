import { SystemModuleExtensionBase } from '../system-module-extension-base';

export class SystemModuleDateExtension extends SystemModuleExtensionBase {
    async date(parsed?: false): Promise<string>;

    async date(parsed?: true): Promise<Date>;

    async date(parsed?: boolean): Promise<string | Date> {
        const { code, stdout } = await this.instance.execute(`date${parsed ? ' -R' : ''}`);
        if (code !== 0) throw new Error('Got non-zero exit code while getting date');
        if (parsed) {
            return new Date(stdout.join(''));
        }
        return stdout.join('');
    }
}
