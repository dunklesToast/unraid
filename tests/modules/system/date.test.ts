import '@system/date';
import { Unraid } from '../../../src';
import { SSHConfig, TestExecutor } from '../../executor/TestExecutor';
import { getInstance } from '../../helpers/getInstance';

describe('system - date', () => {
  let instance: Unraid<SSHConfig, TestExecutor>;

  beforeAll(() => {
    instance = getInstance();
  });

  it('raw', async () => {
    const date = await instance.system.date();
    expect(typeof date).toBe('string');
    expect(date.trim()).toEqual(date);
    expect(date).toMatchSnapshot();
  });

  it('parsed', async () => {
    const date = await instance.system.date(true);
    expect(typeof date).toBe('object');
    expect(date).toMatchSnapshot();
  });
});
