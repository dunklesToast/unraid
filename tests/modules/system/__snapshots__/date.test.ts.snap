// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`system - date parsed 1`] = `Date { NaN }`;

exports[`system - date raw 1`] = `"Sat Dec 18 02:17:50 CET 2021"`;
